banyand/liaison/grpc/node.go

```diff
@@ -0,0 +1,110 @@
+// Licensed to Apache Software Foundation (ASF) under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Apache Software Foundation (ASF) licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package grpc
+
+import (
+	"sync"
+
+	"github.com/pkg/errors"
+
+	databasev1 "github.com/apache/skywalking-banyandb/api/proto/banyandb/database/v1"
+	"github.com/apache/skywalking-banyandb/banyand/metadata"
+	"github.com/apache/skywalking-banyandb/banyand/metadata/schema"
+	"github.com/apache/skywalking-banyandb/pkg/node"
+)
+
+var (
+	_ schema.EventHandler = (*clusterNodeService)(nil)
+	_ NodeRegistry        = (*clusterNodeService)(nil)
+)
+
+// NodeRegistry is for locating data node with group/name of the metadata
+// together with the shardID calculated from the incoming data.
+type NodeRegistry interface {
+	Initialize() error
+	Locate(group, name string, shardID uint32) (string, error)
+}
+
+type clusterNodeService struct {
+	metaRepo metadata.Repo
+	sel      node.Selector
+	sync.Once
+}
+
+// NewClusterNodeRegistry creates a cluster-aware node registry.
+func NewClusterNodeRegistry(metaRepo metadata.Repo, selector node.Selector) NodeRegistry {
```

Comment:
#333 will use `queue.Client` to listen to node events, reducing the need for the liaison node to set up two etcd watchers. Please merge this once PR is merged.